### PR TITLE
Emit `filesystem_changed` only once per frame

### LINKED ISSUE #95979: Slow import with a ton of small files.
### Tested versions

Reproducible in 4.3 stable, 4.4  https://github.com/godotengine/godot/pull/93064

### System information

Sansung galaxy s23+ Forward mobile

### Issue description

This issue might seem related to https://github.com/godotengine/godot/issues/93877 , but not, because the import is alr using vsync dissabled and update continously, so it not that , and also is not the speed itself because importing each files is instant( due to the minimal size in project. 

Anyways , to start with, in a project where you have to import tons of tons of images like over 100k small image files it quite slow, as sure mobile is alrays slower, but taking 2 HOURS to get the import to 50% , is kind of suspicious, as the impprt of each files is instant as most just have small kbs on memory, they are easy to import, and mipmap or compression do not look like they  aggregate time( as it still total instant again, they don't even take 1 second  to import one) , so the cause must be on something else( also note this can be reproducible on  model files tho).

The possible cause: 

An idea i had on what is causing the slow import  might be that is just processing 1 file after another then repeat, when it has the  chance to procced way more at the same time as i said due to importing each being really fast. Another cause might just be the lack of having  the import process in parallel, but i am not  aware if it there in this side or not..

Ways to solve it?

1. Make the reimport process instead of reading  one file at a time, make it proceed like 10 or 20 at the same time as a list and then repeat with other 10  and etc..For it we can use the info we recieved from the mport preparation to classify the files by 10 or 20 as i mentioned. 
2. Make the import process to use threads if not done ( idk if it done alr), ie divide the process by the number of threads and each file reimported has it own thread, or even combine it with what i mentioned of  processing 10 files at the same time without being blocked on just processing 1 by 1 file. 

### Steps to reproduce

- Open the project 

- Wait for everything is scanned until reimport assets popup.

- Notices how the files reading instantly, are still quite slow to import everythimg

### Minimal reproduction project (MRP)

It pretty big so i rediret to it page https://github.com/PMDCollab/SpriteCollab/tree/master

### LINKED ISSUE #97610: Terrible memory leak when importing a large git cloned repository.
### Tested versions

Reproducible in - v4.3.stable.mono.official [77dcf97d8]

### System information

Godot v4.3.stable.mono - Windows 10.0.22631 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2050 (NVIDIA; 32.0.15.5612) - 11th Gen Intel(R) Core(TM) i5-11260H @ 2.60GHz (12 Threads)

### Issue description

A Large Godot Project hosted on a Git server (4 GB of Assets,  >50,000 Files) when cloned and imported without the (.godot) folder, in a machine with 16 GB of RAM will use all system memory and will 100% of the time result in a full system crash, not just the godot Engine. This happens at around 96% to 97% of the import.
Tested on the following Machines
Macbook Pro 15 inch 2017
(Core i7 - 7820HQ, 16GB DDR4 RAM, Radeon Pro 560) (System Crashed) (5 Attempts)
Asus Zepherus G14
(AMD Ryzen 7 6800HS, 16GB DDR5, RTX 3060 ) (BSOD) (4 Attempts)
Asus TUF F15
(Intel Core i7 11260H, 16 GB DDR4, RTX 2050) (System Crashed) (2 Attempts)

Asus TUF F15 (After Upgrade)
(Intel Core i7 11260H, 32 GB DDR4, RTX 2050) (Import Success) (First Attempt)
Desktop 1
(AMD Ryzen 5 5600, 32 GB DDR4, RTX 3070) (Import Success) (First Attempt)
Desktop 2
(AMD Ryzen 5 3600, 32GB DDR4, RTX 2070) (Import Success) (First Attempt)

However even on 32GB of RAM it uses all resources and even after importing, it will continue to use in excess of 22GB RAM
image
but if you restart Godot after importing everything, it will not use more than 2GB of RAM
image

### Steps to reproduce

We cannot unfortunately share steps as it is just importing an existing project

### Minimal reproduction project (MRP)

~~We cannot provide an MRP as the licenses on certain assets do not allow us, how ever we are always available on discord call to talk and maybe help rectify the issue.~~

Equivalent MRP is provide below (Comment 5)

### PR DESCRIPTION
**Pull request by Nazarwadim**
This PR makes it so that only one deferred signal is emitted instead of dozens for each updated file. This was done in order to avoid unnecessary `FileSystemDock::update_all`, which was writing millions of items to the `EditorResourcePreview`
https://github.com/godotengine/godot/blob/61accf060515416da07d913580419fd8c8490f7b/editor/filesystem_dock.cpp#L338-L341
 and creating huge memory usage.

Tested on MRP from  #97610.
Before that, it used 9GB of RAM.
Now it use 1GB.
Also, the import time improved 2-3 times.

Closes #97610
Edit: Does not help #95979

### CODE REVIEW

**Comment by Mickeon** [comment]
Doing this may introduce some unexpected regressions for plugins, as this signal is exposed and has been usable outside of internal code. Although this sounds desirable, it needs to be documented at least.

**Comment by Saul2022** [comment]
Pretty interesting, can’t test rn as i am way too busy, but it could work atleast for ram as my system just has 8gb of ram.

File: editor/editor_file_system.cpp [hunk 1]
```diff
@@ -2376,10 +2376,18 @@ void EditorFileSystem::update_files(const Vector<String> &p_script_paths) {
 		if (!is_scanning()) {
 			_process_update_pending();
 		}
-		call_deferred(SNAME("emit_signal"), "filesystem_changed"); // Update later
+		if (!is_file_system_deferred_updated) {
+			is_file_system_deferred_updated = true;
+			MessageQueue::get_singleton()->push_callable(callable_mp(this, &EditorFileSystem::_update_filesystem)); // Update later.
```
Comment (KoBeWi) [inline editor/editor_file_system.cpp:2381]:
```suggestion
			callable_mp(this, &EditorFileSystem::_update_filesystem).call_deferred(); // Update later.
```

File: editor/editor_file_system.h [hunk 2]
```diff
@@ -189,6 +190,7 @@ class EditorFileSystem : public Node {
 	bool revalidate_import_files = false;
 	int nb_files_total = 0;
 
+	void _update_filesystem();
```
Comment (KoBeWi) [inline editor/editor_file_system.h:193]: I'd call it `_filesystem_changed()`.

File: editor/editor_file_system.h [hunk 3]
```diff
@@ -180,6 +180,7 @@ class EditorFileSystem : public Node {
 	EditorFileSystemDirectory *new_filesystem = nullptr;
 	ScannedDirectory *first_scan_root_dir = nullptr;
 
+	bool is_file_system_deferred_updated = false;
```
Comment (KoBeWi) [inline editor/editor_file_system.h:183]: `filesystem_changed_queued` or something like that.

**Review by KoBeWi — APPROVED** [review]
Looks fine. Usually we have a dedicated method for this pattern, e.g. `queue_redraw()` works like that (i.e. it does a deferred call once per frame). Maybe it's worth applying it for other uses of the `filesystem_changed` signal? But for now this will suffice.

File: editor/editor_file_system.h (hunk 2 above)
Comment (AThousandShips) [reply editor/editor_file_system.h:193]: Or `_notify_filesystem_changed` to be even more clear (`_filesystem_changed` implies it's a handler *connected* to such a signal)

**Comment by Maran23** [comment]
This reminds me on a proposal I saw one time. It proposed to have a way to 'batch' signal emits so that everything that connects to that signal is executed just once per frame. 
Might be worth evaluating as we have this pattern multiple times in the Engine (e.g. LineEdit). And it would be useful for games too.

**Comment by Mickeon** [comment]
A long time ago there was a proposal (or PR?) for a signal connection flag to defer a *single* call, regardless of how many times the signal was emitted (behavior similar to `queue_redraw`). Back in day it was not thought to be useful, but I am very much willing to disagree on that. Just a vague memory.

**Comment by KoBeWi** [comment]
Related:
https://github.com/godotengine/godot-proposals/issues/3245
https://chat.godotengine.org/channel/core?msg=vLiEN4BEM9M8TziDw

**Review by fire — APPROVED** [review]
Wow great find and the improvements on import time is amazing.

**Comment by clayjohn** [comment]
Fantastic work! Thank you!

**Pull request merged**